### PR TITLE
Pump the message queue on frozen windows 

### LINKED ISSUE #16332: Terminal window "not responding"
### Windows Terminal version

1.18.2822.0

### Windows build number

10.0.19045.0

### Other Software

_No response_

### Steps to reproduce

I'm using a terminal window, and then not using it for a while and doing other stuff.

### Expected Behavior

_No response_

### Actual Behavior

The terminal window becomes nonresponsive. The content is gone. The window title shows "not responding" and I can only kill the process at that point.

### PR DESCRIPTION
**Pull request by lhecker**
This changeset ensures that the message queue of frozen windows is
always being serviced. This should ensure that it won't fill up and
lead to deadlocks, freezes, or similar. I've tried _a lot_ of different
approaches before settling on this one. Introducing a custom `WM_APP`
message has the benefit of being the least intrusive to the existing
code base.

The approach that I would have favored the most would be to never
destroy the `AppHost` instance in the first place, as I imagined that
this would be more robust in general and resolve other (rare) bugs.
However, I found that this requires rewriting some substantial parts
of the code base around `AppHost` and it could be something that may
be of interest in the future.

Closes #16332
Depends on #16587 and #16575

### CODE REVIEW

File: src/cascadia/WindowsTerminal/WindowThread.cpp [hunk 1]
```diff
@@ -108,29 +99,22 @@ bool WindowThread::KeepWarm()
...
-        // If ThrowAway() was called, then the buzzer will be signalled without
-        // setting a new _host. In that case, the app is quitting, for real. We
-        // just want to exit with false.
-        const bool reheated = _host != nullptr;
-        if (reheated)
+        if (!GetMessageW(&msg, nullptr, 0, 0))
+        {
+            return false;
+        }
+        if (msg.message == AppHost::WM_REFRIGERATE)
```
Comment (DHowett) [inline src/cascadia/WindowsTerminal/WindowThread.cpp:110]: fwiw mike referred to the act of _waking back up_ as Microwaving; refrigerating is when it goes down to sleep

Comment (DHowett) [reply src/cascadia/WindowsTerminal/WindowThread.cpp:110]: Oh, I see - you have repurposed it to be a single flag. Transition between the states. etc

**Review by DHowett — DISMISSED** [review]

File: src/cascadia/WindowsTerminal/WindowThread.cpp (hunk 1 above)
Comment (lhecker) [reply src/cascadia/WindowsTerminal/WindowThread.cpp:110]: Yeah I wasn't sure what to name it... "refrigerate" has the benefit that it only comes up in this context in the project, so it should be easier to find related code.

File: src/cascadia/WindowsTerminal/AppHost.cpp [hunk 2]
```diff
@@ -535,23 +535,19 @@ void AppHost::LastTabClosed(const winrt::Windows::Foundation::IInspectable& /*se
...
-    // (not by closing the whole window), we need to manually persist an empty
-    // window state here. That will cause the terminal to re-open with the usual
-    // settings (not the persisted state)
-    if (args.ClearPersistedState() &&
-        _windowManager.GetNumberOfPeasants() == 1)
-    {
-        _windowLogic.ClearPersistedWindowState();
-    }
-
     // Remove ourself from the list of peasants so that we aren't included in
```
Comment (zadjii-msft) [inline src/cascadia/WindowsTerminal/AppHost.cpp:538]: Very sus that this is just gone now - why? Is this related?

File: src/cascadia/WindowsTerminal/WindowThread.cpp (hunk 1 above)
Comment (zadjii-msft) [inline src/cascadia/WindowsTerminal/WindowThread.cpp:110]:
```suggestion
        // We're using a single window message (WM_REFRIGERATE) to indicate both 
        // state transitions. In this case, the window is actually being microwaved. 
        if (msg.message == AppHost::WM_REFRIGERATE)
```

File: src/cascadia/WindowsTerminal/WindowThread.cpp [hunk 3]
```diff
@@ -198,6 +182,11 @@ int WindowThread::_messagePump()
 
     while (GetMessageW(&message, nullptr, 0, 0))
     {
+        if (message.message == AppHost::WM_REFRIGERATE)
```
Comment (zadjii-msft) [inline src/cascadia/WindowsTerminal/WindowThread.cpp:185]:
```suggestion
        // We're using a single window message (WM_REFRIGERATE) to indicate both
        // state transitions. In this case, the window is actually being
        // refrigerated. This will break us out of our main message loop, and if
        // we're on windows 10, we'll eventually start the loop in
        // WindowThread::KeepWarm to await being microwaved
        if (message.message == AppHost::WM_REFRIGERATE)
```

File: src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp [hunk 4]
```diff
@@ -26,6 +26,11 @@ NonClientIslandWindow::NonClientIslandWindow(const ElementTheme& requestedTheme)
 {
 }
 
+NonClientIslandWindow::~NonClientIslandWindow()
```
Comment (zadjii-msft) [inline src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp:29]: Just added redundancy? I don't think we've got any code that was removed that was calling `Close` before, right?

**Review by zadjii-msft — COMMENTED** [review]
would approve, but I'm a little curious on the `ClearPersistedWindowState` thing first

File: src/cascadia/WindowsTerminal/AppHost.cpp (hunk 2 above)
Comment (lhecker) [reply src/cascadia/WindowsTerminal/AppHost.cpp:538]: If you expand the surrounding code in the diff viewer you'll notice that this block of code is a duplicate! I think it might have been an incorrect merge in the past.

File: src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp (hunk 4 above)
Comment (lhecker) [reply src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp:29]: Oh, it's terrible... This change fixes a bug and you might already know what I mean when I say: `IslandWindow::Close` is a virtual function which the `~IslandWindow` destructor calls. `NonClientIslandWindow` overloads `Close`. *boom*

But in my opinion this sorely belongs into the "only obvious in hindsight" category of arcane C++ footnukes (can't have enough of those in a language!), so let me explain in more detail:
* No matter whether you define a destructor or not, C++ will define it's own "real-destructor", because that one is responsible for destroying non-trival class members (like `std::string`s and such). So, while `NonClientIslandWindow` doesn't have an explicit destructor, it gets one anyways which virtually overloads the `IslandWindow` destructor.
* With inheritance, destructors are called in reverse starting from child classes up to parent classes. In other words, first `~NonClientIslandWindow` is called and then `~IslandWindow`.

In short, when `~IslandWindow` gets called and it tries to call the virtual function `Close`, the child class `NonClientIslandWindow` is already destroyed, which means that the vtable pointer now points to `IslandWindow`, which means that `IslandWindow::Close` is called and not `NonClientIslandWindow::Close`. The above change fixes this.

**Comment by lhecker** [comment]
FYI this PR sort of depends on #16575 because I had to remove `ThrowAway`. I think it will be fine even if `ExitProcess` is called, but I believe it'll be much safer with `TerminateProcess` so that the UI threads truly don't get a chance to destroy themselves.

**Review by DHowett — APPROVED** [review]

File: src/cascadia/WindowsTerminal/AppHost.cpp (hunk 2 above)
Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/AppHost.cpp:538]: oh what the heck

**Review by zadjii-msft — APPROVED** [review]
